chronosphere/resource_dataset.go

```diff
@@ -0,0 +1,156 @@
+	return &models.ConfigunstableDataset{
+		Configuration: datasetConfigurationToModel(s.Configuration),
+		Description:   s.Description,
+		Name:          s.Name,
+		Slug:          s.Slug,
+	}, nil
+}
+
+func (d datasetConverter) fromModel(m *models.ConfigunstableDataset) (*intschema.Dataset, error) {
+	dsConfig, err := datasetConfigurationFromModel(m.Configuration)
+	if err != nil {
+		return nil, err
+	}
+	return &intschema.Dataset{
+		Name:          m.Name,
+		Slug:          m.Slug,
+		Configuration: dsConfig,
+		Description:   m.Description,
+	}, nil
+}
+
+func datasetConfigurationFromModel(m *models.DatasetDatasetConfiguration) (intschema.DatasetConfiguration, error) {
+	dType := prettyenum.DatasetTypeFromModel(m.Type)
+	switch dType {
+	case prettyenum.DatasetDatasetTypeTracesModel:
+		if m.TraceDataset == nil {
+			return intschema.DatasetConfiguration{}, errors.New("when type = %s, trace_dataset must be provided")
+		}
+		tds, err := traceDatasetFromModel(m.TraceDataset)
+		if err != nil {
+			return intschema.DatasetConfiguration{}, err
+		}
+		return intschema.DatasetConfiguration{
+			Type:         string(dType),
+			TraceDataset: tds,
+		}, nil
+	default:
+		return intschema.DatasetConfiguration{}, errors.Errorf("unsupported type '%s'", dType)
+	}
+}
+
+func datasetConfigurationToModel(s intschema.DatasetConfiguration) *models.DatasetDatasetConfiguration {
+	return &models.DatasetDatasetConfiguration{
+		TraceDataset: traceDatasetToModel(s.TraceDataset),
+		Type:         prettyenum.DatasetType(s.Type).Model(),
+	}
+}
+
+func traceDatasetFromModel(m *models.ConfigunstableTraceDataset) (*intschema.DatasetConfigurationTraceDataset, error) {
+	configV1, err := unstableSearchFilterToV1(m.MatchCriteria)
+	if err != nil {
+		return nil, err
+	}
+	return &intschema.DatasetConfigurationTraceDataset{
+		MatchCriteria: traceSearchFilterFromModel(configV1),
+	}, nil
+}
+
+func traceDatasetToModel(dataset *intschema.DatasetConfigurationTraceDataset) *models.ConfigunstableTraceDataset {
+	if dataset == nil {
+		return nil
+	}
+	configv1Criteria := traceSearchFilterToModel(dataset.MatchCriteria)
+	criteria := v1SearchFilterToUnstable(configv1Criteria)
+	return &models.ConfigunstableTraceDataset{
+		MatchCriteria: criteria,
+	}
+}
+
+func v1SearchFilterToUnstable(criteria *v1models.Configv1TraceSearchFilter) *models.Configv1TraceSearchFilter {
+	if criteria == nil {
+		return nil
+	}
+	serialized, err := criteria.MarshalBinary()
+	if err != nil {
+		return nil
+	}
+	result := &models.Configv1TraceSearchFilter{}
+	err = result.UnmarshalBinary(serialized)
+	if err != nil {
+		return nil
+	}
+	return result
+}
+
+func unstableSearchFilterToV1(criteria *models.Configv1TraceSearchFilter) (*v1models.Configv1TraceSearchFilter, error) {
+	if criteria == nil {
+		return nil, nil
+	}
+	serialized, err := criteria.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	result := &v1models.Configv1TraceSearchFilter{}
+	err = result.UnmarshalBinary(serialized)
```

Comment:
you're sure about this?  Now, we have to make sure that they don't change on the server side (and thus swagger clients) in a way that would break this marshal/unmarshal approach. I think you are sharing a type server-side and this is a generation concern, though, right? In that case it should be fine.

Reply:
> you are sharing a type server-side and this is a generation concern, though, right

this is exactly the reason. I don't love it but this seems to be an artifact of codegen and having an unstable message type import a v1 type, and the seeming 'nestedness' of packages/types in swagger output